app/uk/gov/hmrc/transitmovementseisstub/controllers/MessagesController.scala

```diff
@@ -146,4 +149,33 @@ class MessagesController @Inject() (appConfig: AppConfig, cc: ControllerComponen
       case _                => Some(s"Expected date in RFC 7231 format, instead got $value")
     }
 
+  private def validateLRN(requestBody: Source[ByteString, _]): Either[String, Unit] = {
+
+    val sinkFold: Sink[ByteString, Future[String]] = Sink.fold("") {
+      case (acc, str) =>
+        acc + str.decodeString("UTF-8")
+    }
+
+    val result = requestBody
+      .runWith(sinkFold)
+      .map {
+        string =>
+          loadString(string)
+      }
+      .map {
+        xml =>
+          val lrn = (xml \ "CC015C" \ "TransitOperation" \ "LRN").text
+          lrnRegexPattern
+            .findFirstMatchIn(lrn) match {
+            case Some(_) => Left(s"The supplied LRN: $lrn has already been used by submitter: messageSender")
+            case _       => Right(())
+          }
+      }
+      .recover {
+        case NonFatal(ex) => Right(())
+      }
+
+    Await.result(result, 5.seconds)
```

Comment:
I'm quite nervous about this, there are a couple of things that jump out at me:

* Large XML files, say, getting to 5-8MB, will need to be parsed into a string, then parsed into an XML model, which could take up a lot of memory. Can we use Alpakka and get the LRN out via streaming, if possible? An example of an extractor is at https://github.com/hmrc/transit-movements-router/blob/790435c84b96d00e5032e1a15581bbf2c7faf23c/app/uk/gov/hmrc/transitmovementsrouter/services/XmlParser.scala#L27-L33
* `Await` isn't the best thing here because we're blocking a thread unnecessarily. I guess the best way to handle this is to turn the methods that return `Either`s to return `EitherT`s with `Future`, and then let the runtime handle the futures properly.

Reply:
Yes updated